### ISSUE TITLE
feature: hash source with AST

### ISSUE THREAD

**Issue by majidaldo** [joblib/joblib#1074]
It's a bit annoying that the cache gets invalidated by trivial non-functional code changes. Has anyone thought of hashing the ast to identify a function?

**Comment by ogrisel** [comment]
That's an interesting suggestion. I wonder if it could have a performance impact though.

@pierreglaser wants to investigate with a bad interaction that prevent caching to work as expected when:

- the function is dynamically defined in `__main__`
- the function is cached with memory
- the function is called by Parallel with the loky backend

all 3 at the same time.

Using the AST instead of the source might be a fix for the above problem.

**Comment by majidaldo** [comment]
> * the function is cached with memory
you mean the Memory _object_?

**Comment by pierreglaser** [comment]
I agree that it's a good suggestion. That could also simplify the codebase.
@ogrisel if you think that could be an option, I can investigate this.

**Comment by ogrisel** [comment]
@pierreglaser feel free to investigate, in particular in the context of the bug you described to me IRL and that I tried to summarize in https://github.com/joblib/joblib/issues/1074#issuecomment-652883529. Feel free to open a dedicated issue for the bug itself if you can come up with a simple reproducer.

**Comment by majidaldo** [comment]
I went ahead and gave it a shot. I only did a quick test and it seems to work.

It's crude though. The ast hashing was straightforward but I couldn't follow all the (existing) code. The function source seems to be entangled with its hashing.

**Comment by majidaldo** [comment]
bump @pierreglaser
